### PR TITLE
Git-Trees API

### PR DESCRIPTION
**Pull request by Kasi-R**
I somehow managed to butcher the PR at #4185 

This is an updated PR with a couple changes. 
1. correct git route. (/trees -> /git/trees)
2. additional commenting. 
3. Changed the types used as per https://github.com/go-gitea/go-sdk/pull/94 (GitTreeEntry -> GitEntry) - superficial naming changes only.

Edit: Requires the PR at https://github.com/go-gitea/git/pull/123

### CODE REVIEW

**Comment by lunny** [comment]
should fix #1978

**Comment by lafriks** [comment]
go-gitea/git#123 merged, please update dep

File: routers/api/v1/api.go [hunk 1]
```diff
@@ -433,17 +433,16 @@ func RegisterRoutes(m *macaron.Macaron) {
...
 		m.Post("/org/:org/repos", reqToken(), bind(api.CreateRepoOption{}), repo.CreateOrgRepo)
 
-		m.Group("/repos", func() {
-			m.Get("/search", repo.Search)
-		})
-
 		m.Combo("/repositories/:id", reqToken()).Get(repo.GetByID)
 
 		m.Group("/repos", func() {
+			m.Get("/search", repo.Search)
```
Comment (lunny) [inline routers/api/v1/api.go]: This change should not be in this PR. Maybe put it a separate PR?

**Review by lunny — DISMISSED** [review]

**Review by adelowo — COMMENTED** [review]
Nitpick ... but a lot of variable definitions here seem to use Pascal case instead of camel case as the rest of the project

File: routers/api/v1/repo/tree.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+		return tree
+	}
+	RepoIDLen := len(RepoID)
+
+	// 51 is len(sha1) + len("/git/blobs/"). 40 + 11.
+	BlobURL := make([]byte, RepoIDLen + 51)
+	copy(BlobURL[:], RepoID)
+	copy(BlobURL[RepoIDLen:], "/git/blobs/")
+
+	// 51 is len(sha1) + len("/git/trees/"). 40 + 11.
```
[the remark this answers is not in the document]
Comment (lafriks) [reply routers/api/v1/repo/tree.go:58]: I don't think it is worth to add it as constant

File: routers/api/v1/repo/tree.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	"code.gitea.io/sdk/gitea"
+)
+
+func GetTree(ctx *context.APIContext) {
+	sha := ctx.Params("sha")
+	if len(sha) == 0 {
+		ctx.Error(400, "sha not provided", nil)
+		return
+	}
+	Tree := GetTreeBySHA(ctx, sha)
```
Comment (lafriks) [inline routers/api/v1/repo/tree.go]: I agree with @adelowo, please lowercase first letter for all local variables

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/api/v1/api.go (hunk 1 above)
Comment (Kasi-R) [reply routers/api/v1/api.go]: I removed that change from this commit. Will submit a new PR for it.

File: routers/api/v1/repo/tree.go (hunk 3 above)
Comment (Kasi-R) [reply routers/api/v1/repo/tree.go]: Done.

Comment (adelowo) [reply routers/api/v1/repo/tree.go]: https://github.com/go-gitea/gitea/pull/5403/files#diff-f0fb94da81e656bdc93dabff1e5077d2R23

Comment (Kasi-R) [reply routers/api/v1/repo/tree.go]: Whoops. Thanks!

File: routers/api/v1/repo/tree.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	"fmt"
+	"strings"
+
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/git"
+	"code.gitea.io/sdk/gitea"
+)
+
+func GetTree(ctx *context.APIContext) {
```
Comment (techknowlogick) [inline routers/api/v1/repo/tree.go:18]: This function needs a comment preceding it explaining what it does.

Exact words from CI are:
```
/go/src/code.gitea.io/gitea/routers/api/v1/repo/tree.go:17:1: exported function GetTree should have comment or be unexported
```

File: routers/api/v1/repo/tree.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	tree := GetTreeBySHA(ctx, sha)
+	if tree != nil {
+		ctx.JSON(200, tree)
+	} else {
+		ctx.Error(400, "sha invalid", nil)
+	}
+}
+
+func GetTreeBySHA(ctx *context.APIContext, sha string) *gitea.GitTreeResponse {
```
Comment (techknowlogick) [inline routers/api/v1/repo/tree.go:33]: Same as above

**Review by techknowlogick — CHANGES_REQUESTED** [review]

**Review by techknowlogick — APPROVED** [review]

**Comment by techknowlogick** [comment]
@lunny @lafriks please update review

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/api/v1/api.go [hunk 6]
```diff
@@ -480,6 +480,9 @@ func RegisterRoutes(m *macaron.Macaron) {
 			m.Group("/:username/:reponame", func() {
 				m.Combo("").Get(reqAnyRepoReader(), repo.Get).
 					Delete(reqToken(), reqOwner(), repo.Delete)
+				m.Group("/git/trees", func() {
```
[the remark this answers is not in the document]
Comment (techknowlogick) [reply routers/api/v1/api.go]: updated

**Review by lafriks — APPROVED** [review]

**Comment by richmahn** [comment]
@Kasi-R Were there any API docs added for this, or plans to be added? https://try.gitea.io/api/v1/swagger doesn't show anything for git/trees

**Comment by sapk** [comment]
The docs seems to be missing in this PR and are  done in a separate one https://github.com/go-gitea/gitea/pull/5834
